### PR TITLE
Runtime: define get_return_address for WebAssembly

### PR DESCRIPTION
**Pull request by MaxDesiatov**
WebAssembly/WASI doesn't support stack traces and return addresses. To make it possible to compile exclusivity code for this platform, it seems to make the most sense for the `get_return_address()` to return a zero value. Since it's only used for diagnostics, this shouldn't cause any issues when targeting WASI.

Related to SR-9307.

(cc @compnerd)

### CODE REVIEW

File: stdlib/public/runtime/Exclusivity.cpp [hunk 1]
```diff
@@ -38,14 +38,22 @@
...
+#elif defined(__wasm__)
+#define get_return_address() ((void*) 0)
 #else
 #error missing implementation for get_return_address
 #define get_return_address() ((void*) 0)
 #endif
 
 using namespace swift;
 
+#if defined(__wasm__)
```
Comment (compnerd) [inline stdlib/public/runtime/Exclusivity.cpp:50]: Should this not be `__wasi__`, because this is not related to the object file format?

**Comment by compnerd** [comment]
CC: @gottesmm @mikeash

**Comment by compnerd** [comment]
Id like to get @gottesmm's opinion on this as well

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Comment by atrick** [comment]
What does exclusivity checking have to do with threads? Other than the implementation needing to be thread-safe, which is trivial if there are no threads.

**Comment by MaxDesiatov** [comment]
Great point, sorry I missed that. I've re-enabled it, now the only thing left in this PR is defining `get_return_address`.

**Comment by MaxDesiatov** [comment]
@compnerd could you have another look please?

**Review by compnerd — COMMENTED** [review]
Does this not break other things as you cannot get the return address?  This may let it compile but can cause correctness issues.

**Comment by karwa** [comment]
@compnerd @MaxDesiatov

Just looking in to this. It seems that [WASI does not support](https://github.com/WebAssembly/WASI/issues/159) getting the return address right now. AFAICT this is only used for diagnostics, so it shouldn’t be a big deal to leave it out.

The other option (besides returning a null pointer), would be to omit the “PC” field from the “Access” type on WASI. Would you prefer that approach?

**Comment by MaxDesiatov** [comment]
@compnerd `get_return_address()` is only used in `SwiftTLSContext::get().accessSet.insert` call, which as far as I understand exists for diagnostics. In my understanding the worst that can happen is that `AccessSet` will have a zero value in the `pc` field. That isn't dereferenced, and is only used for diagnostic output.

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
I think that it may be better to disable the larger portions of Swift for WASM until the get return address can be implemented.  Note that the else path would still prevent you from progressing, which this is undoing.

**Comment by kateinoigakukun** [comment]
Close due to the merge of https://github.com/apple/swift/pull/42097
